### PR TITLE
Fix parsing of java codegen module prefixes

### PR DESCRIPTION
**Pull request by cocreature**
I admit I have no idea what I thought when I wrote that split but it
was clearly nothing sensible.

changelog_begin

- [Java Codegen] Fix parsing of `module-prefixes` to support package
  names containing dashes.

changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: language-support/codegen-common/src/test/scala/com/digitalasset/daml/lf/codegen/conf/CodegenConfigReaderSpec.scala [hunk 1]
```diff
@@ -114,6 +115,91 @@ class CodegenConfigReaderSpec extends AnyFlatSpec with Matchers {
...
+        .map(x => PackageReference.NameVersion(name, PackageVersion.assertFromString(x)) -> prefix)
+        .toMap,
+    )
+
+    codegenConf(badConfigStr, Java) shouldBe Right(
+      expected
+    )
+  }
+
+  it should "reject package references with no dash" in {
```
Comment (stefanobaghino-da) [inline language-support/codegen-common/src/test/scala/com/digitalasset/daml/lf/codegen/conf/CodegenConfigReaderSpec.scala:186]: Can we reliably make this assumption? Should we (can we?) think about separating the library name and version?

File: language-support/codegen-common/src/main/scala/com/digitalasset/daml/lf/codegen/conf/CodegenConfigReader.scala [hunk 2]
```diff
@@ -157,13 +157,15 @@ object CodegenConfigReader {
 
   implicit val decodePackageReference: KeyDecoder[PackageReference] =
     new KeyDecoder[PackageReference] {
+      // This regex matches IdString.PackageVersion.
+      private val refRegex = """^(.*)-((0|[1-9][0-9]*)(\.(0|[1-9][0-9]*))*)$""".r
```
Comment (stefanobaghino-da) [inline language-support/codegen-common/src/main/scala/com/digitalasset/daml/lf/codegen/conf/CodegenConfigReader.scala:161]: This worries me a bit. It also statically assumes a specific format for the version number, which we probably don't want us to tie ourselves to. Given the code that you edited below (unless the comment I made applies), it appears that we can assume that the last component separated by a `-` is the version. Can we do that instead?

**Review by stefanobaghino-da — COMMENTED** [review]
Thanks for the fix. I'm not 100% convinced by the solution though.

File: language-support/codegen-common/src/main/scala/com/digitalasset/daml/lf/codegen/conf/CodegenConfigReader.scala (hunk 2 above)
Comment (cocreature) [reply language-support/codegen-common/src/main/scala/com/digitalasset/daml/lf/codegen/conf/CodegenConfigReader.scala:161]: This is specified in daml-lf so we _need_ to tie ourselves to that.

File: language-support/codegen-common/src/test/scala/com/digitalasset/daml/lf/codegen/conf/CodegenConfigReaderSpec.scala (hunk 1 above)
Comment (cocreature) [reply language-support/codegen-common/src/test/scala/com/digitalasset/daml/lf/codegen/conf/CodegenConfigReaderSpec.scala:186]: Yes we can safely make this assumption because GHC makes this assumption and we cannot compile a package that does not follow this pattern. If we ever move away from that we could do something else but that doesn’t seem worth considering atm.

File: language-support/codegen-common/src/main/scala/com/digitalasset/daml/lf/codegen/conf/CodegenConfigReader.scala (hunk 2 above)
Comment (cocreature) [reply language-support/codegen-common/src/main/scala/com/digitalasset/daml/lf/codegen/conf/CodegenConfigReader.scala]: I might have misunderstood your point. I switched to splitting on the last dash which doesn’t tie us any less to the format (PackageVersion.fromString will explode) but it removes the duplication which seem snice I guess

File: language-support/codegen-common/src/main/scala/com/digitalasset/daml/lf/codegen/conf/CodegenConfigReader.scala [hunk 3]
```diff
@@ -158,17 +158,20 @@ object CodegenConfigReader {
...
+          None
+        } else {
           for {
-            name <- PackageName.fromString(name).toOption
-            version <- PackageVersion.fromString(version).toOption
+            name <- PackageName.fromString(key.slice(0, dashIndex)).toOption
+            version <- PackageVersion.fromString(key.slice(dashIndex + 1, key.length)).toOption
           } yield PackageReference.NameVersion(name, version)
-        case _ => None
+        }
```
Comment (stefanobaghino-da) [inline language-support/codegen-common/src/main/scala/com/digitalasset/daml/lf/codegen/conf/CodegenConfigReader.scala:174]: I don't understand the need for a private method that is just invoked, you can probably do something like the following:
```
    key => {
      key.splitAt(key.lastIndexOf('-')) match {
        case (rawPackageName, rawPackageVersion)
            if rawPackageName.nonEmpty && rawPackageVersion.nonEmpty =>
          for {
            name <- PackageName.fromString(rawPackageName).toOption
            version <- PackageVersion.fromString(rawPackageVersion).toOption
          } yield PackageReference.NameVersion(name, version)
        case _ =>
          None
      }
```

Comment (cocreature) [reply language-support/codegen-common/src/main/scala/com/digitalasset/daml/lf/codegen/conf/CodegenConfigReader.scala]: Inlined it, I don’t think your suggestion works. You are not dropping the dash in between. I also prefer handling the fact where there is no dash explicitly instead of relying on PackageName.fromString to blow up on an empty name (which you get for splitting at -1).

Comment (stefanobaghino-da) [reply language-support/codegen-common/src/main/scala/com/digitalasset/daml/lf/codegen/conf/CodegenConfigReader.scala]: > relying on PackageName.fromString to blow up on an empty name (which you get for splitting at -1)

I'm not sure I follow this, the guard clearly prevents you from going into that branch if either is an empty string.

Comment (cocreature) [reply language-support/codegen-common/src/main/scala/com/digitalasset/daml/lf/codegen/conf/CodegenConfigReader.scala]: Oh sorry right I missed that guard. You still have the issue that you’re not dropping the dash.

Comment (cocreature) [reply language-support/codegen-common/src/main/scala/com/digitalasset/daml/lf/codegen/conf/CodegenConfigReader.scala]: switched to something closer to what you suggested, I find that harder to follow but your choice.

Comment (stefanobaghino-da) [reply language-support/codegen-common/src/main/scala/com/digitalasset/daml/lf/codegen/conf/CodegenConfigReader.scala]: Yes, I agree that that's hard to follow. I didn't want you to go in that direction but I don't want you to bug any further, just merge as is, I'll change it afterwards.

Comment (cocreature) [reply language-support/codegen-common/src/main/scala/com/digitalasset/daml/lf/codegen/conf/CodegenConfigReader.scala]: apologies for being a bit slow in following what you were suggesting, I’ll merge as you suggested and let you clean it up. Sorry for the extra work caused by that.

Comment (stefanobaghino-da) [reply language-support/codegen-common/src/main/scala/com/digitalasset/daml/lf/codegen/conf/CodegenConfigReader.scala]: By the way, your code was silently relying on `PackageName.fromString` to blow up on an empty name or version (i.e. a dash at the beginning or end of the string).

Comment (cocreature) [reply language-support/codegen-common/src/main/scala/com/digitalasset/daml/lf/codegen/conf/CodegenConfigReader.scala]: good point

Comment (stefanobaghino-da) [reply language-support/codegen-common/src/main/scala/com/digitalasset/daml/lf/codegen/conf/CodegenConfigReader.scala]: Took care of it in https://github.com/digital-asset/daml/pull/12982
